lib/segment/src/index/hnsw_index/graph_links.rs

```diff
@@ -69,62 +144,151 @@ impl GraphLinks {
+            let links_mmap: &mut [PointOffsetType] = transmute_from_u8_mut(links_mmap);
+            let offsets_mmap: &mut [u64] = transmute_from_u8_mut(offsets_mmap);
+            offsets_mmap[0] = 0;
+
+            let mut links_pos = 0;
+            let mut offsets_pos = 1;
+            for level in 0..header.levels_count as usize {
+                level_offsets.push(offsets_pos as u64 - 1);
+                self.iterate_level_points(level, |_, links| {
+                    links_mmap[links_pos..links_pos + links.len()].copy_from_slice(links);
+                    links_pos += links.len();
+
+                    offsets_mmap[offsets_pos] = links_pos as u64;
+                    offsets_pos += 1;
+                });
             }
-            result.push(layers);
         }
-        result
+
+        {
+            let level_offsets_range = header.get_level_offsets_range();
+            let level_offsets_byte_slice = &mut mmap[level_offsets_range];
+            let level_offsets_slice: &mut [u64] = transmute_from_u8_mut(level_offsets_byte_slice);
+            level_offsets_slice.copy_from_slice(&level_offsets);
+        }
+
+        mmap.flush()?;
+        Ok(())
+    }
+
+    pub fn get_levels_count(&self) -> usize {
+        // because back_index is sorted by point`s max layer, we can retrieve max level from `point_id = back_index[0]`
+        self.edges[self.back_index[0]].len()
     }
 
-    pub fn links(&self, point_id: PointOffsetType, level: usize) -> &[PointOffsetType] {
+    pub fn iterate_level_points<F>(&mut self, level: usize, mut f: F)
+    where
+        F: FnMut(usize, &mut Vec<PointOffsetType>),
+    {
+        let edges_len = self.edges.len();
         if level == 0 {
-            self.get_links_at_zero_level(point_id)
+            (0..edges_len).for_each(|point_id| f(point_id, &mut self.edges[point_id][0]));
         } else {
-            self.get_links_at_nonzero_level(point_id, level)
+            for i in 0..edges_len {
+                let point_id = self.back_index[i];
+                if level >= self.edges[point_id].len() {
+                    break;
+                }
+                f(point_id, &mut self.edges[point_id][level]);
+            }
         }
     }
+}
 
-    pub fn num_points(&self) -> usize {
-        self.reindex.len()
+pub trait GraphLinks: Default {
+    fn load_from_file(path: &Path) -> OperationResult<Self>;
+
+    fn offsets_len(&self) -> usize;
+
+    fn levels_count(&self) -> usize;
+
+    fn get_links(&self, range: Range<usize>) -> &[PointOffsetType];
+
+    fn get_links_range(&self, idx: usize) -> Range<usize>;
+
+    fn get_level_offset(&self, level: usize) -> usize;
+
+    fn reindex(&self, point_id: PointOffsetType) -> PointOffsetType;
+
+    fn num_points(&self) -> usize;
+
+    // Convert from graph layers builder links
+    // `Vec<Vec<Vec<_>>>` means:
+    // vector of points -> vector of layers for specific point -> vector of links for specific point and layer
+    fn from_vec(
```

Comment:
`from_vec` have an unexpected side-effect of saving data to the file. More over, this side-effect does not work with mmap version. Could be better to avoid mixing functionalities.

Reply:
That's by design because mmap mode should avoid making a full copy of links and writes directly to disk. I propose to change name to `initialize`